gstd/src/critical.rs

```diff
@@ -0,0 +1,130 @@
+// This file is part of Gear.
+
+// Copyright (C) 2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Critical section that guarantees code section execution
+//!
+//! Code is executed in `handle_signal` entry point in case of failure
+//! only across [`wait`](crate::exec::wait) or `.await` calls
+//! because sections have to be saved.
+//!
+//! ```rust,no_run
+//! use gstd::{critical, msg};
+//!
+//! # async fn _dummy() {
+//!
+//! // get source outside of critical section
+//! // because `gr_source` sys-call is forbidden inside `handle_signal` entry point
+//! let source = msg::source();
+//! // register section
+//! let section = critical::Section::new(move || {
+//!     msg::send(source, "example", 0).expect("Failed to send message");
+//! });
+//!
+//! // section is now saved
+//! msg::send_for_reply(msg::source(), "for_reply", 0, 0)
+//!     .expect("Failed to send message")
+//!     .await
+//!     .expect("Received error reply");
+//!
+//! // if some code fails (panic, out of gas, etc) after `wait` (`send_for_reply` in our case)
+//! // then saved sections will be executed in `handle_signal`
+//!
+//! // your code
+//! // ...
+//!
+//! // execute section
+//! section.execute();
+//!
+//! # }
+//! ```
+
+use alloc::boxed::Box;
+use core::{any::TypeId, mem};
+use hashbrown::HashMap;
+
+static mut SECTIONS: Option<Sections> = None;
+
+pub(crate) struct Sections {
+    fns: HashMap<TypeId, Box<dyn FnMut()>>,
```

Comment:
If we will have two sections to the same function, whether this will not break logic in case one is dropped:
```rust
fn func() {}

let s1 = Section::new(func);

{
     let _ = Section::new(func);
     // in drop remove section from table by `func` TypeId
}

something.await;

panic!(); // whether this panic cause `s1` execution in signal handler?

```